### ISSUE TITLE
Some possible fixes.

### ISSUE THREAD

**Issue by ammasakshay** [joker-ware/JokerCord#2]
Took me a while, but I managed to get it to work (despite never using python before lol)
anyway here is my 2 cents
— Some names are wrong. You likely know this and are fixing it as you yourself mention this.
— It catches in EVERY server. This makes it dangerous, and easy to get banned. Perhaps add an option to restrict it to specific servers/channels
—Lists. Supposing you don't want to catch every pokemon, perhaps it would be nice to create a list and it only catches those specific pokemon. This option should also be allowed to work in all servers or the current server, as you may only want to spam catch in your present server but if a legend appears somewhere else you want to nab it.
— Optional catch delay. Add a random delay on some catches, when using on other servers so it's not obvious you are botting. Delay specified by the user.

Thanks for an amazing app. Great job!

**Comment by ammasakshay** [comment]
- Another feature that would enhance the experience is the log. Sometimes you run the program overnight, and obviously a lot of pokemon will be caught. Rather than using the unintuitive interface of pokecord to try and figure out what is new, it would be very nice to have a log showing the caught pokemon over a time period— Building on this, it would be even nicer if it was sorted into normal, rare and legendary pokemon. So you will wake up and instantly know if you caught a legendary.

**Comment by joker-ware** [comment]
I really really appreciate your feedback! Thank you for taking your time in reviewing my work. I will get into it as soon as possible, this not was a personal project and as you might have seen it still is in a very early stage. Those suggestions are very useful for me. Thank you very much again for this kind of feedback, it feels nice to see that your job is being acknowledged :D

**Comment by ammasakshay** [comment]
No problem. Always glad to help. Especially since it's so wonderful.  Another comment is your read-me. It instructs you to use the bash command to run it. For people like me who have no idea what you are talking about, it took me nearly 40 minutes as well as almost installing linux, before I realized that I could just double click on the file to execute it. (I run windows by the way)

**Comment by joker-ware** [comment]
A beginner friendly readme has always been in my head. I will try to explain it as simple as possible in the next update.

**Comment by joker-ware** [comment]
Did some updates, still haven't finished. You can wait until the next update or try this one out.

**Comment by joker-ware** [comment]
Also making the begginer guide, just finished Linux's. Will get into Windows and Mac when I have some time

**Comment by ammasakshay** [comment]
I will try it out.

**Comment by joker-ware** [comment]
Found a critical bug. Please wait until the next update.

**Comment by joker-ware** [comment]
Should have been fixed. This version is stable-

**Comment by ammasakshay** [comment]
I tried the new update. 
Here are the comments
— Pokemon identification still needs work. 
— Can you add an option to add a slight delay between pokemon and catch? Pokecord notices if a pokemon is caught within a second of it spawning as it usually takes a few seconds to type. It will ignore it if done once or twice, but if it is done consistantly then you are likely to get banned. What I have seen in other bots is they add a priority list. The pokemon on the lists get caught instantly and the rest after a small random delay to make it look realistic.
— I like the pokemon list option, but it's manual and time-consuming to add it. A better way is to simply create a list .txt file named list with one pokemon name on each line. So you just edit the .txtt file to make your list. Then the UI can ask you are you using a list (True/False) and if True name the list, to which you respond (list name).txt
—Multiple servers should be possible. What you can do is again use a loop. Catch on all servers (True/False). If False enter Channel Id. Once this is done the questing is Enter another Channel ID (True/False) If False then ends. If True keep repeating till all channels. Or maybe a more user friendly version of that LOL.
Hope it helps

**Comment by joker-ware** [comment]
Thanks for your reply. I'm trying my best to fix pokemon names but that's something that will require huge research from my side and I currently don't have that much free time. However, it is my main priority right now, but sure it will be fixed as soon as I find a solution. Talking about the delays, I will implement it in this update as it is something that will not require much work and don't know why I forgot to do so. And yes, I know the manual list is a struggle but it was a temporary solution and I'm not at all discarding the option to actually add a Graphical User Interface. On the other hand multiple servers are also gonna be possible in the next update, just locked it to the spam channel right away as atemporary solution to prevent bans.

Again, thank you for your time and contributions to this project, you don't know how happy that makes me as a developer that just started in this world. :)

**Comment by joker-ware** [comment]
Currently working on a GUI that will manage everything related to the bot. Will update soon

**Comment by joker-ware** [comment]
The web interface is nearly done after a lot of work. Soon I'll be updating JokerCord to 0.0.4 and with the GUI, which will include many (but not all) of the ideas you sent. Make sure to test it when it releases ! :)

**Comment by ammasakshay** [comment]
I will do that for sure!. Will the update also help in pokemon identification?

**Comment by joker-ware** [comment]
Sadly not. After this update I will prioritize my time on fixing it, but as I said I don't have a clue on how to do so.

**Comment by ammasakshay** [comment]
I can't really give you a suggestion here, because I'm not sure how yours gets names, BUT I'm assuming its something like this

STEP 1 - Copy the image that the script detects whenever pokecord spawns it.
STEP 2 - Match it 
Step 3- Input p!catch <name>

How I've seen other bots do STEP 2 is, that they include in the download a set of jpegs of the entire Pokedex. Then they just match the spawned picture to the Pokedex picture (that can be included in in the download of your program)

![pokedex](https://user-images.githubusercontent.com/48739144/66385288-ddd0be80-e9dd-11e9-9449-c354ec2109bc.PNG)

**Comment by joker-ware** [comment]
Actually to minimize the folder size, all the Pokemon's are contained inside /Assets/phash.json and dhash.json. They are premade hashes of every pokemon in the pokedex (which I get automatically with an unitily I created) and the script just hashes the Pokemon Pokecord sends as it receives them and compares it with the database. The problem is that Pokecors messes up a bit with their sent images because as far as I have seem, it sends .jpegs when they actually are pngs, so it messes up with my script. I don't know why the efficiency is so low as they images are the same, and my final guess is the .jpeg Pokecord sends. They must do that in purpose.

**Comment by ammasakshay** [comment]
Do you really have to minimise folder size though? If making the user download the entire Pokedex and use it removes errors, I think you should stick with that. Unless you can outsmart the pokecord people and fix it.

**Comment by joker-ware** [comment]
Not only the folder size is minimized but the processing speed. There should be no difference between hashing the pokemon in the moment to having it hashed, if so, hashes would actually be more precise, imo

**Comment by joker-ware** [comment]
New version is out! Make sure to install the new library on readme and follow the instructions

**Comment by joker-ware** [comment]
Amazingly, I think I just found how to fix the detection. Imma work on it.

**Comment by joker-ware** [comment]
It actually was simpler than I thought...

**Comment by ammasakshay** [comment]
I'm actually trying to figure out how to run the new version. It's immediately closing after opening. Also, the BASH command doesnt exist on windows. You should probably take it out of your windows tutorial.

**Comment by ammasakshay** [comment]
Alright, it works now. I opened it in an IDLE shell it gave the error FLASK module not found. After pip3 install flask it ran.

**Comment by ammasakshay** [comment]
Ok. The UI needs work. 
- The reboot button doesn't work. At least not on windows. I had to end up restarting it manually. The caught list is wonderful. Perhaps add an option, Don't catch duplicate? Or Release Duplicate with lower ID. (In later ones)
- The list function is nice. However manual pokemon by pokemon entry is hard. Perhaps give a few convenient options, like catch legendary, catch ultra beast etc. 
-Edit preference is wonky. For example, after typing everything in and hitting save, it looks like this. It still works when I restart it, but the UI itself leaves you wondering if it worked or not.
-Channel. Ok, this isn't exactly what I meant. I wouldn't fuse spam channel and catch channel. Maybe what you can do is have another list which each line being a channel that you want pokecord to operate in. This stops restricting you in one channel.
Otherwise, great work. It looks amazing, and can't believe it has grown this much!
![bug](https://user-images.githubusercontent.com/48739144/66448919-d73f5700-ea70-11e9-9a34-97297324b0fb.PNG)
Good Luck! and thanks for your work!

**Comment by ammasakshay** [comment]
Maybe what you can do for the edit preference is when you select auto spam false, then it doesn't ask you for the other two on the UI.

**Comment by joker-ware** [comment]
Thanks for your feedback! Some points:
- The reboot button worked when I tested it. I will take a look at it when I can
- That caught option is nice! Looking forward to adding it.
- I like the idea of adding legendary pokemons. This only was a temporary placeholder for a more complex custom list page.
- I actually did notice while coding it that after saving the Edit preferences it showed like that, but if shouldn't have given you an error. Did you fill all of the information (even the interval and text, if you did not want auto spam?) 
- the channel option is just temporary as I work on fixing the multi server catching. Don't worry it won't stay like this.

Thank you again! Also, I will push the auto catch fix today, it should be much better.

**Comment by GiGaMo24** [comment]
Just adding some info for you about the reboot function.
It is working for me. but whenever i refresh the page, it would asking for confirmed submission. I just click it but the page didnt load and in the terminal it show that the reboot message again.

**Comment by joker-ware** [comment]
Yes, I forgot. Until I fix it, please **do not** resend submission, otherwise it will reboot the program again and again. Just go to the search bar and hit enter or type http://localhost:5555

**Comment by ammasakshay** [comment]
For the edit preference what happens is this.
Possibility 1
You fill in all the information
Token -  xxxx
Channel ID - xxx
Spam - Yes/No
Spam Intervel --xx
Spam text --xx
**It works.** However when you reopen the edit, on restarting Spam interval and Spam text become blank and DO NOT save your preferences.
Possibility 2
Channel ID - xxx
Spam - Yes/No
Here you select NO (for the do you want auto spam option). If you do not fill in the rest it will return an error. My suggestion was made it so that it does not ask for all the information if you select no for auto spam (this is purely UI not a bug or anything). Also that in possibility 1, it saves your preferences for auto spam and spam text in the box.

A much more complex suggestion is this. Allow multiple lists in order of priority. List one legendaryies/mythicals/ultrabeasts etc. These are all caught instantly. Others can be caught after a small delay, to make it feel like a human.
As for lists, my suggestion is of course to leave the add custom list option but also to have some pre made lists.
-Legendaries
-Mythicals
-Ultra Beasts
-Pseudo-Legendaries (https://bulbapedia.bulbagarden.net/wiki/Pseudo-legendary_Pok%C3%A9mon)
- Starters
-Everything
-Custom List
Each list can be set on different levels of priority. This means I can have legendaries on high priority but everything on low priority which means legends will be insta caught and the remaining will be caught after a delay or occasionally ignored. Second, add the dont catch duplicate function here and allow it to apply per list I'd say. This allows me to get multiple of the same legendary, (if it does happen because legendaries are super rare) and not, say, get 1000 weedles.
Hope this helped!

**Comment by joker-ware** [comment]
@ammasakshay Did the pip3 command work with &&? I changed it recently because I thought just doing 
```bash
pip3 install lib1 lib2...
```
Would work on Windows. Please confirm

**Comment by ammasakshay** [comment]
I did it manually lol
each command one by one
so without the and. Also, you need to mention the add Flask module. That is something that is needed past this update.

**Comment by joker-ware** [comment]
Two days ago I ran JokerCord on a Windows VM and doing pip3 install discord.py && pip3 install dhash...etc
Worked for me.

**Comment by ammasakshay** [comment]
It might work. I was just saying that I did it manually (one at a time)

**Comment by joker-ware** [comment]
Pokemon detection has been fixed completely. Let me now of you find any error.

**Comment by ammasakshay** [comment]
Have caught 40 pokemon without error. I don't expect to find one and it is likely fixed. Congrats!

**Comment by ammasakshay** [comment]
It is excellent and easily usable. Id say the bot is no longer in beta and is fully functioning. No further fixes are all enhancement.
I would say the order of priority would be something like this (Maybe)
-> Channels. Give full control of where and how this is used. Can it be used in a single channel. Can it be used in multiple channels. Can it be used only in specific channels in specific servers. Can you restrict spam to a single channel, yet also catch everywhere?
After this update, you have full functionality.
-> Lists. I have already mentioned it before, but this would likely be next in priority. Some premade lists (Legendaries/mythical/ultra beasts/pseudo legendaries)
->Catch-delay options. To leave a realistic delay for non-priority captures. for example, you can select any list as a priority and still attempt to catch everything. This will show other users that you were too slow to catch some, and yet you will snag what you want. Probably will reduce you being banned
-> Minor tweaks in the GUI (It still forces you to fill out spam interval and spam content even if you select don't spam in edit preferences.
I can't think of any other cool features yet, but I will add to this when I can. 
Great Work!

**Comment by joker-ware** [comment]
Actually, I'd better update the channels functionality before the next update.

**Comment by joker-ware** [comment]
Finally I added the option to enable or disable guilds on which to catch pokemons. You can modify it in the new version on the webpage. (Settings page)

**Comment by ammasakshay** [comment]
quick fix. Dont add gender to catches. for example, in pokecord, both nidoran male and female are caught by p!catch nidoran. you dont need p!catch Nidoran♂. Anyone who sees that will almost certainly suspect you are botting. you also dont need to capitalise the first letter of the pokemon. These are things people typically dont do when speed typing.
-This is NOT a bug. This is an enhancement that will reduce people getting caught if they use this bot.

**Comment by joker-ware** [comment]
Okay, will change it ASAP

**Comment by joker-ware** [comment]
Nidoran changed. ^^

**Comment by joker-ware** [comment]
Okay, Pokemons are now sent lowercase.

**Comment by ammasakshay** [comment]
Awesome. Will keep trying to think of new features. It's really awesome and already more functional than most pokecord bots out there.

**Comment by joker-ware** [comment]
Okay, added the option to set a custom catch delay for each server. Can access it from settings.

**Comment by prosoxamit** [comment]
All i need is your help, I am new to python, I have installed Python 3.7 and after running joker.py and navigating to http://localhost:5555 from Chrome, it doesn't work. I am kinda new to Python, can you help me what to do after installing Python 3.7.
Waiting for you reply :)

**Comment by prosoxamit** [comment]
xD sorry, i didnt read the Readme, now its working fine, thanks :P

**Comment by joker-ware** [comment]
Going to close this until you find any more improvements or I release a new update :)
